### ISSUE TITLE
[Feature Request] Layer Normalization

### ISSUE THREAD

**Issue by Kaixhin** [pytorch/pytorch#1959]
See https://github.com/pytorch/pytorch/issues/1601 for previous discussion on [layer normalization](https://arxiv.org/abs/1607.06450).

**Comment by jekbradbury** [comment]
I use this:

```python
class LayerNorm(nn.Module):

    def __init__(self, features, eps=1e-6):
        super().__init__()
        self.gamma = nn.Parameter(torch.ones(features))
        self.beta = nn.Parameter(torch.zeros(features))
        self.eps = eps

    def forward(self, x):
        mean = x.mean(-1, keepdim=True)
        std = x.std(-1, keepdim=True)
        return self.gamma * (x - mean) / (std + self.eps) + self.beta
```

**Comment by Kaixhin** [comment]
Nice @jekbradbury - though I presume 2D versions etc. would have to be adapted to work on a per-channel basis, like batch norm? Either way, it's still probably worth having this in the library with a test.

**Comment by jekbradbury** [comment]
Yeah, as usual I'm more or less forgetting that images exist. This should at least generalize to a couple different situations in NLP though, or really anywhere that channels are the last dimension and you want separate moments for every {timestep, y-position, etc.}.

**Comment by weihancug** [comment]
This is the Layer normalization [implementation ](https://github.com/igul222/improved_wgan_training/blob/master/tflib/ops/layernorm.py)in tensorflow. So How I can transfer it to pytorch implementation , how to transfer the nn.moments and etc..

`

1
--
import tflib as lib
  |  
  | import numpy as np
  | import tensorflow as tf
  |  
  | def Layernorm(name, norm_axes, inputs):
  | mean, var = tf.nn.moments(inputs, norm_axes, keep_dims=True)
  |  
  | # Assume the 'neurons' axis is the first of norm_axes. This is the case for fully-connected and BCHW conv layers.
  | n_neurons = inputs.get_shape().as_list()[norm_axes[0]]
  |  
  | offset = lib.param(name+'.offset', np.zeros(n_neurons, dtype='float32'))
  | scale = lib.param(name+'.scale', np.ones(n_neurons, dtype='float32'))
  |  
  | # Add broadcasting dims to offset and scale (e.g. BCHW conv data)
  | offset = tf.reshape(offset, [-1] + [1 for i in xrange(len(norm_axes)-1)])
  | scale = tf.reshape(scale, [-1] + [1 for i in xrange(len(norm_axes)-1)])
  |  
  | result = tf.nn.batch_normalization(inputs, mean, var, offset, scale, 1e-5)
  |  
  | return result

`

**Comment by weihancug** [comment]
This is the Layer normalization [implementation](https://github.com/igul222/improved_wgan_training/blob/master/tflib/ops/layernorm.py) in tensorflow. So How I can transfer it to pytorch implementation , how to transfer the nn.moments and etc..

```python
def Layernorm(name, norm_axes, inputs):
  mean, var = tf.nn.moments(inputs, norm_axes, keep_dims=True)
  # Assume the 'neurons' axis is the first of norm_axes. This is the case for fully-connected and BCHW conv layers.
  n_neurons = inputs.get_shape().as_list()[norm_axes[0]]
   
  offset = lib.param(name+'.offset', np.zeros(n_neurons, dtype='float32'))
  scale = lib.param(name+'.scale', np.ones(n_neurons, dtype='float32'))
   
  # Add broadcasting dims to offset and scale (e.g. BCHW conv data)
  offset = tf.reshape(offset, [-1] + [1 for i in xrange(len(norm_axes)-1)])
  scale = tf.reshape(scale, [-1] + [1 for i in xrange(len(norm_axes)-1)])
  
  result = tf.nn.batch_normalization(inputs, mean, var, offset, scale, 1e-5)
    
  return result
```

**Comment by 1nadequacy** [comment]
I have created a naive implementation of LayerNorm for GRU recently. It makes things quite a bit slower, mainly because everything is being done manually, but hopefully it can be helpful.

Also note, that I don't divide hidden state by std, because I usually initialize the hidden state with zeros. 

```python
class LayerNormGRUCell(nn.GRUCell):
    def __init__(self, input_size, hidden_size, bias=True):
        super(LayerNormGRUCell, self).__init__(input_size, hidden_size, bias)

        self.gamma_ih = nn.Parameter(torch.ones(3 * self.hidden_size))
        self.gamma_hh = nn.Parameter(torch.ones(3 * self.hidden_size))
        self.eps = 0

    def _layer_norm_x(self, x, g, b):
        mean = x.mean(1).expand_as(x)
        std = x.std(1).expand_as(x)
        return g.expand_as(x) * ((x - mean) / (std + self.eps)) + b.expand_as(x)

    def _layer_norm_h(self, x, g, b):
        mean = x.mean(1).expand_as(x)
        return g.expand_as(x) * (x - mean) + b.expand_as(x)

    def forward(self, x, h):

        ih_rz = self._layer_norm_x(
            torch.mm(x, self.weight_ih.narrow(0, 0, 2 * self.hidden_size).transpose(0, 1)),
            self.gamma_ih.narrow(0, 0, 2 * self.hidden_size),
            self.bias_ih.narrow(0, 0, 2 * self.hidden_size))

        hh_rz = self._layer_norm_h(
            torch.mm(h, self.weight_hh.narrow(0, 0, 2 * self.hidden_size).transpose(0, 1)),
            self.gamma_hh.narrow(0, 0, 2 * self.hidden_size),
            self.bias_hh.narrow(0, 0, 2 * self.hidden_size))

        rz = torch.sigmoid(ih_rz + hh_rz)
        r = rz.narrow(1, 0, self.hidden_size)
        z = rz.narrow(1, self.hidden_size, self.hidden_size)

        ih_n = self._layer_norm_x(
            torch.mm(x, self.weight_ih.narrow(0, 2 * self.hidden_size, self.hidden_size).transpose(0, 1)),
            self.gamma_ih.narrow(0, 2 * self.hidden_size, self.hidden_size),
            self.bias_ih.narrow(0, 2 * self.hidden_size, self.hidden_size))

        hh_n = self._layer_norm_h(
            torch.mm(h, self.weight_hh.narrow(0, 2 * self.hidden_size, self.hidden_size).transpose(0, 1)),
            self.gamma_hh.narrow(0, 2 * self.hidden_size, self.hidden_size),
            self.bias_hh.narrow(0, 2 * self.hidden_size, self.hidden_size))

        n = torch.tanh(ih_n + r * hh_n)
        h = (1 - z) * n + z * h
        return h

class LayerNormGRU(nn.Module):
    def __init__(self, input_size, hidden_size, bias=True):
        super(LayerNormGRU, self).__init__()
        self.cell = LayerNormGRUCell(input_size, hidden_size, bias)
        self.weight_ih_l0 = self.cell.weight_ih
        self.weight_hh_l0 = self.cell.weight_hh
        self.bias_ih_l0 = self.cell.bias_ih
        self.bias_hh_l0 = self.cell.bias_hh

    def forward(self, xs, h):
        h = h.squeeze(0)
        ys = []
        for i in range(xs.size(0)):
            x = xs.narrow(0, i, 1).squeeze(0)
            h = self.cell(x, h)
            ys.append(h.unsqueeze(0))
        y = torch.cat(ys, 0)
        h = h.unsqueeze(0)
        return y, h
```

**Comment by D-X-Y** [comment]
@jekbradbury Hi, Jek, is there any more efficient implementation of LayerNorm, I use your code, it is very cool, but as there are more than 50 LN in my network, I feel it becomes the bottemneck of the speed.

**Comment by LynnHo** [comment]
@jekbradbury, this is the LayerNorm for >=2D modified from your code

```python
class LayerNorm(nn.Module):

    def __init__(self, num_features, eps=1e-5, affine=True):
        super(LayerNorm, self).__init__()
        self.num_features = num_features
        self.affine = affine
        self.eps = eps

        if self.affine:
            self.gamma = nn.Parameter(torch.Tensor(num_features).uniform_())
            self.beta = nn.Parameter(torch.zeros(num_features))

    def forward(self, x):
        shape = [-1] + [1] * (x.dim() - 1)
        mean = x.view(x.size(0), -1).mean(1).view(*shape)
        std = x.view(x.size(0), -1).std(1).view(*shape)

        y = (x - mean) / (std + self.eps)
        if self.affine:
            shape = [1, -1] + [1] * (x.dim() - 2)
            y = self.gamma.view(*shape) * y + self.beta.view(*shape)
        return y
```

**Comment by calclavia** [comment]
@D-X-Y @jekbradbury I can also confirm that a simple implementation of Layernorm seems to be quite slow on Pytorch. Maybe the mean and standard deviation computation could be combined into a single CUDA kernel?

**Comment by t-vi** [comment]
Just a quick note: @LynnHo 's version calculates the mean and std over all channels combined. If you want per-channel mean and std, you could use the following similar code.

Best regards

Thomas

```python
class LayerNorm(torch.nn.Module):
    def __init__(self, features, affine=True, eps=1e-6):
        super().__init__()
        self.affine = affine
        self.eps = eps
        if self.affine:
            self.gamma = torch.nn.Parameter(torch.ones(features))
            self.beta = torch.nn.Parameter(torch.zeros(features))
    def forward(self, x):
        additionaldims = (x.dim()-2)*[1]
        batch_size = x.size(0)
        channels = x.size(1)
        mean = x.view(batch_size, channels, -1).mean(2).view(batch_size, channels, *additionaldims)
        std  = x.view(batch_size, channels, -1).std(2).view(batch_size, channels, *additionaldims)
        x = (x - mean) / (std + self.eps)
        if self.affine:
            x = self.gamma.view(1, channels, *additionaldims) * x + self.beta.view(1, channels, *additionaldims)
        return x
```

**Comment by LynnHo** [comment]
@t-vi I just follow how tf-slim does. However, this implementation is unbearably slow. Anybody know more effective implementation?

**Comment by t-vi** [comment]
There seems to be one for TF, maybe you can use that as inspiration: https://github.com/MycChiu/fast-LayerNorm-TF .

**Comment by Blade6570** [comment]
@LynnHo I am new to Pytorch and just going through the layer-normalization code. The mean and std will return a single value even for images in CNN? Just bit confused

**Comment by grantsrb** [comment]
@Blade6570 according to the [Layer Normalization paper](https://arxiv.org/abs/1607.06450), yes the mean and standard deviation should be a single number shared between all activations (see section 3 of the paper). This is true for all network types.

The paper also notes that Batch Normalization works better than Layer Normalization for Convolutional Neural Networks (see section 6.7).

Note that there is the possibility that new research has been released since the release of this paper (July 21, 2016) that I am unaware of.

**Comment by Kaixhin** [comment]
Closed by https://github.com/pytorch/pytorch/pull/4922 - thanks @SsnL!

**Comment by wandering007** [comment]
Are the implementations of Layer Normalization and [Instance Normalization](http://pytorch.org/docs/0.3.1/nn.html#torch.nn.InstanceNorm1d) identical?

**Comment by SsnL** [comment]
@wandering007 They are very similar, but have some subtle differences. IN is applied on each channel of channeled data like images, but LN is usually applied on entire sample and often in NLP tasks. Also LN applies elementwise affine transform, while IN usually don't apply affine transform. PyTorch does support IN with a scalar affine transform applied to each channel, but that's mostly coming from that IN is implemented with BN code.

**Comment by wandering007** [comment]
@SsnL Thanks for your patient explanation. So if I want to use LN in the context of channeled data like images, I should use IN instead? e.g. ops are convolution instead of linear.

**Comment by SsnL** [comment]
You can use both on images. They just work differently. Depending on your design and purpose of the work, I can see either (and with either affine on or off) work.

**Comment by wandering007** [comment]
@SsnL Thanks, I want to use normalization to improve ConvLSTM performance for video classification. Is IN with affine on more suitable?

**Comment by faroit** [comment]
@SsnL your explanation here was super helpful to me. I would suggest to add a bit more documentation in the IN layer description to clarify this.

**Comment by SsnL** [comment]
@faroit Adding in https://github.com/pytorch/pytorch/pull/11106
